src/transformers/trainer.py

```diff
@@ -2491,7 +2490,13 @@ def _inner_training_loop(
                     if step % args.gradient_accumulation_steps == 0:
                         self.control = self.callback_handler.on_step_begin(args, self.state, self.control)
 
-                    with self.accelerator.accumulate(model):
+                    # We explicitly want to avoid relying on `accelerator.accumulate` for generation training
+                    context = (
+                        functools.partial(self.accelerator.no_sync, model=model)
+                        if i == len(batch_samples) - 1
+                        else contextlib.nullcontext
+                    )
```

Comment:
For an explanation on what we have going on here @Rocketknight1 , during DDP we use `model.no_sync()` to only communicate across all GPUs during the next step outside it (so we speed up training when not needed when doing gradient accumulation). `accelerator.no_sync()` is the lower-level `accumulate()` API which makes that op backed-independent (so on a single GPU it just does `nullcontext`)